File: GamePad-Mobile/pc-server/Protocol/Enums.cs
namespace GamePadServer.Protocol;

/// <summary>
/// Message types used in the GamePad binary protocol.
/// </summary>
public enum MessageType : byte
{
    ButtonEvent = 0x01,
    Discovery = 0x02,
    DiscoveryResponse = 0x03,
    AxisEvent = 0x04,
    Vibration = 0x05,
    MacroEvent = 0x06,
    Ping = 0x07,
    Pong = 0x08,
    TriggerEvent = 0x09,
}

/// <summary>
/// Identifiers for every physical/virtual button on the controller.
/// </summary>
public enum ButtonId : byte
{
    A = 0x01,
    B = 0x02,
    X = 0x03,
    Y = 0x04,
    LB = 0x05,
    RB = 0x06,
    LT = 0x07,
    RT = 0x08,
    Start = 0x09,
    Select = 0x0A,
    Guide = 0x0B,
    DPadUp = 0x0C,
    DPadDown = 0x0D,
    DPadLeft = 0x0E,
    DPadRight = 0x0F,
    L3 = 0x10,
    R3 = 0x11,
}

/// <summary>
/// Binary state of a button.
/// </summary>
public enum ButtonState : byte
{
    Released = 0x00,
    Pressed = 0x01,
    LongPressed = 0x02,
}

/// <summary>
/// Identifies which analog stick generated an axis event.
/// </summary>
public enum StickId : byte
{
    Left = 0x01,
    Right = 0x02,
}

/// <summary>
/// Identifies an analog trigger.
/// </summary>
public enum TriggerId : byte
{
    L2 = 0x01,
    R2 = 0x02,
}

File: GamePad-Mobile/mobile/lib/core/protocol/enums.dart
/// Message types used in the GamePad binary protocol.
enum MessageType {
  buttonEvent(0x01),
  discovery(0x02),
  discoveryResponse(0x03),
  axisEvent(0x04),
  vibration(0x05),
  macroEvent(0x06),
  ping(0x07),
  pong(0x08),
  triggerEvent(0x09);

  final int value;
  const MessageType(this.value);

  static MessageType fromValue(int value) {
    return MessageType.values.firstWhere(
      (e) => e.value == value,
      orElse: () => MessageType.buttonEvent,
    );
  }
}

/// Identifiers for every physical/virtual button on the controller.
enum ButtonId {
  a(0x01),
  b(0x02),
  x(0x03),
  y(0x04),
  lb(0x05),
  rb(0x06),
  lt(0x07),
  rt(0x08),
  start(0x09),
  select(0x0A),
  guide(0x0B),
  dPadUp(0x0C),
  dPadDown(0x0D),
  dPadLeft(0x0E),
  dPadRight(0x0F),
  l3(0x10),
  r3(0x11);

  final int value;
  const ButtonId(this.value);
}

/// Binary state of a button.
enum ButtonState {
  released(0x00),
  pressed(0x01),
  longPressed(0x02);

  final int value;
  const ButtonState(this.value);
}

/// Identifies which analog stick.
enum StickId {
  left(0x01),
  right(0x02);

  final int value;
  const StickId(this.value);
}

/// Identifies an analog trigger.
enum TriggerId {
  l2(0x01),
  r2(0x02);

  final int value;
  const TriggerId(this.value);
}
